src/main/java/org/apache/commons/imaging/ImageParser.java

```diff
@@ -79,25 +78,18 @@
  * that the documentation is perfect, especially in the more obscure
  * and specialized areas of implementation.
  *
- * <h3>The "Map params" argument</h3>
+ * <h3>The "params" argument</h3>
  *
- * Many of the methods specified by this class accept an argument of
- * type Map giving a list of parameters to be used when processing an
- * image. For example, some of the output formats permit the specification
+ * <p>Many of the methods specified by this class accept an argument of
+ * type {@code T} defining the parameters to be used when
+ * processing an image. For example, some of the output formats permit
  * of different kinds of image compression or color models. Some of the
  * reading methods permit the calling application to require strict
- * format compliance.   In many cases, however, an application will not
- * require the use of this argument.  While some of the ImageParser
- * implementations check for (and ignore) null arguments for this parameter,
- * not all of them do (at least not at the time these notes were written).
- * Therefore, a prudent programmer will always supply an valid, though
- * empty instance of a Map implementation when calling such methods.
- * Generally, the java HashMap class is useful for this purpose.
+ * format compliance.</p>
  *
- * <p>Additionally, developers creating or enhancing classes derived
- * from ImageParser are encouraged to include such checks in their code.
+ * @param <T> type of parameters used by this image parser
  */
-public abstract class ImageParser extends BinaryFileParser {
+public abstract class ImageParser<T extends ImagingParameters> extends BinaryFileParser {
```

Comment:
Would it make sense to remove this generic parameter and only accept `ImagingParameters` in methods? Implementations could downcast as needed to access type-specific parameters and this would simplify usage where the image format is not known (as in `Imaging`).

Reply:
I remember trying that.. now I'm in the process of trying that again, and I think I've hit the same wall as in the last time.

The `ImageParser` has some methods like

```java
public abstract BufferedImage getBufferedImage(ByteSource byteSource, ImagingParameters params)
            throws ImageReadException, IOException;
```

Then in `BmpImageParser`, for example, when I try

```java
    @Override
    public BufferedImage getBufferedImage(final ByteSource byteSource, final BmpImagingParameters params)
            throws ImageReadException, IOException {
```

It fails due to the `Override` annotation, with `Method does not override method from its superclass`. If I rename the parameter to `ImagingParameter`, that compiles. But then we could have users sending a `JpegImagingParameters` instead, and it would fail in runtime.

I think this is where I stopped last time, and thought I would have to either accept `ImagingParameters` everywhere, or use generics... any advice @darkma773r ?

Last item pending in this review :tada: 

Thanks!!!
Bruno

Reply:
Here's an idea for a plan (or at least 12 percent of an plan):

In the package example.common, we have the class ExampleParameters and the interface ExampleParser

```
public class ExampleParameters {
    public int test;
}
public interface ExampleParser {
    public void parse(ExampleParameters paramaters);
}
```

There are multiple implementing format packages.  In package example.format.alpha, we have

```
public class AlphaParameters extends ExampleParameters {
    int mess;
    /**
     * A copy constructor
     * @param source 
     */
    AlphaParameters(ExampleParameters source){
        this.mess =source.test;
    }
}

public class AlphaParser implements ExampleParser {
    @Override
    public void parse(ExampleParameters parameters) {
          AlphaParameters alphaParameters;
          if(parameters instanceof AlphaParameters){
               alphaParameters = (AlphaParameters)parameters;
          }else{
               alphaParameters = new AlphaParameters(parameters);
         }
          parse(alphaParameters);
    }
    
    public void parse(AlphaParameters parameters){
        // do the work
    }
}

```

My thinking is that most applications that have format-specific parameters, will call their parser class directly and pass in the specific class that goes with it (in this example, AlphaParameter... in real life TiffImagingParameters, etc.).   Even those that passed in a generic would experience a small overhead for the downcast or the copy constructor...  Those application that are use the high-level ImageParser class (which determines format and branches based on format type) would pass in an instance of the base class ExampleParameters.  After all, it's only a small performance hit and if they really have format-specific needs (as I do in my applications), they would instantiate a format-specific parser.

This approach has the advantage of avoiding Generics, which some developers might find challenging or obscure, especially in this particular case. 

Also, the copy constructor that I used in this example wouldn't always be necessary.    A parser class could also do something like:
```
     @Override
    public void parse(ExampleParameters parameters) {
     if (parameters instanceof AlphaParameters) {
            processAlphaParameters((AlphaParameters) parameters);
        } else {
            processGeneralParameters(parameters);
        }
```

This approach has one disadvantage in that we couldn't test for all contingencies at compile time.  For example, if we had 

```
       // BetaParameters extends ExampleParameters
        BetaParameters betaParameters = new BetaParameters();  
        AlphaParser alphaParser = new AlphaParser();
        alphaParser.parse(betaParameters);
```

This would compile just fine.  However, betaParameters would contain information that was irrelevant to the alpha parser. It would probably be a case where the programmer made a mistake.   We could introduce run-time checking into all of our parsers, or we could just let it go. 

I'm leaning to the "just let it go option", but here's one way that the run-time checking could be accomplished.  In the AlphaParser class, we have

```
        if (parameters instanceof AlphaParameters) {
            alphaParameters = (AlphaParameters) parameters;
        } else {
            enforceUseOfBaseClass(parameters);
            alphaParameters = new AlphaParameters(parameters);
        }
```

And the method:

```
   void enforceUseOfBaseClass(ExampleParameters parameters){
        if(!ExampleParameters.class.equals(parameters.getClass())){
            throw new IllegalArgumentException("parameters is not the base class");
        }
    }
```

Reply:
Maybe forget the whole bit about the copy constructor for AlphaParameters.  I think most classes would just do the branching approach and only the guys who needed to do something special would use the copy-constructor approach.

```
     @Override
    public void parse(ExampleParameters parameters) {
     if (parameters instanceof AlphaParameters) {
            processAlphaParameters((AlphaParameters) parameters);
        } else {
            processGeneralParameters(parameters);
        }
```

Reply:
Hi @gwlucastrig 

An approach similar to what we are trying here, thanks for sharing! 

>This would compile just fine. However, betaParameters would contain information that was irrelevant to the alpha parser. It would probably be a case where the programmer made a mistake. The program would probably run, but it might not do quite what the developer hoped. Some debugging would be necessary. How could we help this case? 

I think we have reached the same point here. While both our approaches lead to no generics, the cost is accepting compilation and having to do some runtime checking, though not perfect.

>Maybe forget the whole bit about the copy constructor for AlphaParameters. I think most classes would just do the branching approach and only the guys who needed to do something special would use the copy-constructor approach.

My current experiment is not going anywhere, so I'll `git reset --hard` and start testing yours with the copy constructor.

Reply:
@gwlucastrig, your approach is almost exactly like the one I was going to suggest :-)
- Have `ImageParser` and other classes accept `ImagingParameters` directly.
- Add format-specific parameter overloads of these methods as needed,
- Have the general methods either cast the general `ImagingParameters` to the format-specific ones or use a copy constructor to extract relevant fields while leaving others as defaults.

> I'm leaning to the "just let it go option", but here's one way that the run-time checking could be accomplished.

I'm also leaning toward the "just let it go option" for users that pass format-specific parameters to parsers of the wrong format type. In my mind, any format-specific parameters that do not apply should just be ignored. I'm guessing this was the case when the API used a `Map` to pass parameters so we shouldn't be losing anything. I would also not expect a method that accepts an `ImagingParameters` argument to succeed if I pass one parameter subclass but not another one. 

Side note: @kinow, I'm looking at the `ImagingParameters` class and I can't find any usages of the `getImageFormat()` method elsewhere in the code. Is that method used anywhere?

Reply:
>I'm also leaning toward the "just let it go option" for users that pass format-specific parameters to parsers of the wrong format type. In my mind, any format-specific parameters that do not apply should just be ignored.

Yeah, I realize that this would be simpler for us to maintain, but I still think this could lead to problems to users. There certain formats that people may be confused about which parameter class to use.

For example, someone could try the TIFF for JPEG, or vice-versa. The `TiffImagingParameters` extends `XmpImagingParameters`. And the `JpegImagingParameters` extends the `TiffImagingParameters`. So you could have code using a Tiff parser with the `JpegImagingParameters` that works fine. Until we - for instance - override a property like compression in the JPEG parameter. That could break the code of the user, and would require some debugging to understand why that started happening.

Maybe we should consult with others in the mailing list?

>Side note: @kinow, I'm looking at the ImagingParameters class and I can't find any usages of the getImageFormat() method elsewhere in the code. Is that method used anywhere?

Ah! It's not! Brilliant catch @darkma773r ! The parsers are using the format to write/read. And a parser knows its own format. No need to have it in the parameters used in the parser. Thanks a lot for finding this one @darkma773r , less one new public API that would need updating later on :+1: 

Removing it.

Reply:
Well, I am thinking back to my reaction when Java 1.5 first introduced generics.  My first thought was "what a good idea".  My second thought was "of all the mistakes I have made coding in Java (and there are a lot of them), I don't think I ever tried to store the wrong kind of object in a collection".   My point here is that we need to simplify things for the user as much as possible, but if our own code starts to get too complicated we might be going down the wrong road.  Also, I wonder why JpegImagingParameters should extend TiffImagingParameters. The TIFF format has MANY Tiff-unique elements (support for old fashioned FAX parameters, Photogrammetric Interpreters, odd-duck compression standards).  I would think that if there were any overlap between the two package branches at all, perhaps JpegImagingParameters and TiffImagingParameters might have a common ancestor.   Of course, I have been so completely focused on TIFF that I haven't really studied the JPEG package. So I am not sure how it is organized.

Reply:
Good point @gwlucastrig . I will have a look to remember why I extended the tiff class.

As for the generics, maybe we can find a compromise between our two options, even if it requires further changes (e.g. rethink the Imaging class for example). Or maybe after reviewing this jpeg - tiff parent relationship we may simplify this code more and have some other ideas.
Thanks!!!

Reply:
A couple more thoughts here: 
- I like the idea of having the TIFF and JPEG parameters extend a common ancestor if possible. That seems like a more maintainable solution. With the current hierarchy, if a property needed to be added to the TIFF parameters, it would also be added to the JPEG one, regardless of whether or not it applied to that format.
- With the removal of the `imageFormat` property, we could completely remove all of the `ImagingParameters` subclasses that do not add any properties, such as `BmpImagingParameters`. They do not seem to add much value anymore and could easily be added later if needed.

Reply:
Separated TIFF and JPEG parameters, now they share a common ancestor, but JPEG doesn't extend TIFF.